httpbp/errors.go

```diff
@@ -0,0 +1,560 @@
+// ErrorResponse.WithTemplateName.
+func (r ErrorResponse) TemplateName() string {
+	if r.templateName != "" {
+		return r.templateName
+	}
+	return DefaultErrorTemplateName
+}
+
+// NewErrorResponse returns a new ErrorResponse with the given inputs.
+//
+// This can be used to create custom NewErrorResponses, but it is encouraged that
+// developers use the standard ones provided by httpbp rather than creating them
+// directly.
+func NewErrorResponse(code int, reason string, explanation string) *ErrorResponse {
+	return &ErrorResponse{
+		code:        code,
+		Reason:      reason,
+		Explanation: explanation,
+		Details:     make(map[string]string),
+	}
+}
+
+var (
+	_ HTMLBody     = ErrorResponse{}
+	_ HTMLBody     = (*ErrorResponse)(nil)
+	_ fmt.Stringer = ErrorResponse{}
+	_ fmt.Stringer = (*ErrorResponse)(nil)
+)
+
+// JSONError returns the given error as an HTTPError that will write JSON.
+func JSONError(resp *ErrorResponse, cause error) HTTPError {
+	body := ErrorResponseJSONWrapper{Success: false, Error: resp}
+	return newHTTPError(resp.code, body, cause, JSONContentWriter())
+}
+
+// HTMLError returns the given error as an HTTPError that will write HTML.
+func HTMLError(resp *ErrorResponse, cause error, t *template.Template) HTTPError {
+	return newHTTPError(resp.code, resp, cause, HTMLContentWriter(t))
+}
+
+// RawError returns the given error as an HTTPError that will write a raw
+// response of the given content-type.
+func RawError(resp *ErrorResponse, cause error, contentType string) HTTPError {
+	return newHTTPError(resp.code, resp, cause, RawContentWriter(contentType))
+}
+
+// RegisterDefaultErrorTemplate adds the default HTML template for error pages to the
+// given templates and returns the result.
+//
+// This only registeres the single, shared, default template, if you want to
+// use custom HTML templates for specific errors, you will need to customize the
+// template name on the error by calling `SetTemplateName`
+func RegisterDefaultErrorTemplate(t *template.Template) (*template.Template, error) {
+	return t.New(DefaultErrorTemplateName).Parse(ErrorPageTemplate)
+}
+
+// RegisterCustomDefaultErrorTemplate adds the custom template passed in as the
+// default HTML template for error pages, rather than the default template
+// provided by baseplate.
+//
+// If you are creating a custom template, note that the ErrorResponse struct is
+// what is passed to the template to render so you are limited to the values
+// exported by ErrorResponse.
+func RegisterCustomDefaultErrorTemplate(t *template.Template, text string) (*template.Template, error) {
+	return t.New(DefaultErrorTemplateName).Parse(text)
+}
+
+// BadRequest is for 400 responses.
+//
+// This is appropriate for when the client sends a malformed or invalid request.
+//
+// If the client sent an invalid request, you can send the details using the
+// Details map in the ErrorResponse.
+func BadRequest() *ErrorResponse {
```

Comment:
🔕 Do you think it makes sense to construct a map of http code -> *ErrorResponse? We can also provide a function instead, fallback to a generic one (maybe the 500 one) if it's not in the map.

Reply:
We can't just _directly_ map code to *ErrorResponse because some of the methods mutate the error response, so we either need to update the existing code to clone the ErrorResponse in those methods and return a new ErrorResponse, or have a function that takes in the http code and returns a new *ErrorResponse.

Also, to clarify, are you suggesting that rather than having a dedicated function for each error code, we have a single one that takes in an http error code?  I think that the ergonomics of that are probably _worse_ than the current approach.  Or maybe this map is hidden and all the these functions do is return the appropriate error from that map?

Reply:
It doesn't have to be a map of code -> response, it could be from code to a function to give you a response. (or a function that takes in code as arg and return a function)

Reply:
Hmm, maybe have both options? You the give me a code, you get back an error can be used in cases where someone has multiple potential errors and wants to easily pick one while the majority of cases use the specific functions directly?

It also might help to give these functions a standard prefix, say `Err` so it's `httpbp.ErrForbidden`, etc?

Reply:
yea give both options makes sense. as for common prefix, I don't think `Err` is a good one (it's more commonly used by pre defined error variables), and I can't think of a good one so maybe not?